presto-common/src/main/java/com/facebook/presto/common/block/DictionaryBlock.java

```diff
@@ -264,14 +264,29 @@ public long getRegionLogicalSizeInBytes(int positionOffset, int length)
         }
 
         long sizeInBytes = 0;
-        long[] seenSizes = new long[dictionary.getPositionCount()];
-        Arrays.fill(seenSizes, -1L);
-        for (int i = positionOffset; i < positionOffset + length; i++) {
-            int position = getId(i);
-            if (seenSizes[position] < 0) {
-                seenSizes[position] = dictionary.getRegionLogicalSizeInBytes(position, 1);
+        // Dictionary Block may contain large number of keys and small number of indices.
+        // This happens during the file read or write, where the dictionary of the file
+        // is much larger than number of rows read. In these cases, compute the size
+        // in place and do not use the cache code path. The assumption is, creating a cache,
+        // cache lookup, and cache population is equal to the cost of method call.
+        // The conversion ratio can be tuned if required to add a constant factor on either side.
+        if (length > dictionary.getPositionCount()) {
```

Comment:
So the idea is that it doesn't make sense to create a seenSizes cache if the cached seenSize value is not going to be reused, right?

Reply:
That is another way of looking at it, and that comment probably makes this more simpler. 

I was looking at it from the opinion, that there is a cost to call getRegionLogicalSizeInBytes on a block. There is a cost to do the cache maintenance.